go/tasks/plugins/presto/execution_state.go

```diff
@@ -0,0 +1,528 @@
+	extraArgs := currentState.CurrentPrestoQuery.ExtraArgs
+
+	response, err := prestoClient.ExecuteCommand(ctx, statement, extraArgs)
+	if err != nil {
+		// If we failed, we'll keep the NotStarted state
+		currentState.CreationFailureCount = currentState.CreationFailureCount + 1
+		logger.Warnf(ctx, "Error creating Presto query for %s, failure counts %d. Error: %s", uniqueID, currentState.CreationFailureCount, err)
+	} else {
+		executeResponse := response.(client.PrestoExecuteResponse)
+
+		// If we succeed, then store the command id returned from Presto, and update our state. Also, add to the
+		// AutoRefreshCache so we start getting updates for its status.
+		commandID := executeResponse.ID
+		logger.Infof(ctx, "Created Presto ID [%s] for token %s", commandID, uniqueID)
+		currentState.CommandID = commandID
+		currentState.Phase = PhaseSubmitted
+		currentState.URI = executeResponse.NextURI
+
+		executionStateCacheItem := ExecutionStateCacheItem{
+			ExecutionState: currentState,
+			Identifier:     uniqueID,
+		}
+
+		// The first time we put it in the cache, we know it won't have succeeded so we don't need to look at it
+		_, err := cache.GetOrCreate(uniqueID, executionStateCacheItem)
+		if err != nil {
+			// This means that our cache has fundamentally broken... return a system error
+			logger.Errorf(ctx, "Cache failed to GetOrCreate for execution [%s] cache key [%s], owner [%s]. Error %s",
+				tCtx.TaskExecutionMetadata().GetTaskExecutionID().GetID(), uniqueID,
+				tCtx.TaskExecutionMetadata().GetOwnerReference(), err)
+			return currentState, err
+		}
+	}
+
+	return currentState, nil
+}
+
+func MonitorQuery(
+	ctx context.Context,
+	tCtx core.TaskExecutionContext,
+	currentState ExecutionState,
+	cache cache.AutoRefresh) (ExecutionState, error) {
+
+	uniqueID := tCtx.TaskExecutionMetadata().GetTaskExecutionID().GetGeneratedName()
+	executionStateCacheItem := ExecutionStateCacheItem{
+		ExecutionState: currentState,
+		Identifier:     uniqueID,
+	}
+
+	cachedItem, err := cache.GetOrCreate(uniqueID, executionStateCacheItem)
+	if err != nil {
+		// This means that our cache has fundamentally broken... return a system error
+		logger.Errorf(ctx, "Cache is broken on execution [%s] cache key [%s], owner [%s]. Error %s",
+			tCtx.TaskExecutionMetadata().GetTaskExecutionID().GetID(), uniqueID,
+			tCtx.TaskExecutionMetadata().GetOwnerReference(), err)
+		return currentState, errors.Wrapf(errors.CacheFailed, err, "Error when GetOrCreate while monitoring")
+	}
+
+	cachedExecutionState, ok := cachedItem.(ExecutionStateCacheItem)
+	if !ok {
+		logger.Errorf(ctx, "Error casting cache object into ExecutionState")
+		return currentState, errors.Errorf(errors.CacheFailed, "Failed to cast [%v]", cachedItem)
+	}
+
+	// If there were updates made to the state, we'll have picked them up automatically. Nothing more to do.
+	return cachedExecutionState.ExecutionState, nil
+}
+
+func MapExecutionStateToPhaseInfo(state ExecutionState) core.PhaseInfo {
```

Comment:
does this function still work?  I feel like we need to add more phase versions.

The Hive state machine is a one pass waterfall.  it just goes through each state linearly.  This state machine goes back and revisits the states because there's the query count.  We should make each time around a new phase version.